File: sanitize.go
package dopdb

import (
	"fmt"
	"strings"
)

// ----------------------------------------------------------------------------
// Filter sanitization.
//
// In Redis/doptime the frontend could only invoke a CLOSED set of verbs, so a
// per-key on/off whitelist was a sufficient safety model. FIND reopened that
// surface: a filter is an arbitrary document supplied by the caller.
//
// On KVRocks the filter is no longer handed to a database — query.go evaluates
// it in this process (see the dialect note below). That removes the server-side
// injection class outright, but NOT the need for this function: an unbounded
// filter is still a denial-of-service surface (pathological nesting, operators
// with no bounded cost) and the allowlist is what keeps the accepted dialect
// identical to the one the TypeScript engine implements. Anything outside the
// list is rejected here rather than silently ignored downstream.
//
// The operator names are unchanged from the Mongo build on purpose: the HTTP
// wire protocol, the typed clients and the docs all speak this dialect, and the
// storage swap was not a reason to break them. The evaluator in query.go accepts
// exactly this set.
// ----------------------------------------------------------------------------

// allowedQueryOps are filter operators considered safe to accept from callers.
var allowedQueryOps = map[string]bool{
	// comparison
	"$eq": true, "$ne": true, "$gt": true, "$gte": true,
	"$lt": true, "$lte": true, "$in": true, "$nin": true,
	// logical
	"$and": true, "$or": true, "$nor": true, "$not": true,
	// element
	"$exists": true, "$type": true,
	// array
	"$all": true, "$elemMatch": true, "$size": true,
	// evaluation (safe subset)
	"$regex": true, "$options": true, "$mod": true,
}

// forbiddenOps are operators that executed code, performed writes, or traversed
// collections in the Mongo dialect. query.go implements none of them, so they
// could simply fall through the allowlist — they are named explicitly anyway,
// because a filter that contains one is a caller expecting behaviour dopdb will
// not provide, and a clear error beats a silent empty result set.
var forbiddenOps = map[string]bool{
	"$where": true, "$function": true, "$accumulator": true,
	"$expr":   true, // $expr can embed $function/$let; disallow wholesale
	"$lookup": true, "$graphLookup": true, "$unionWith": true,
	"$merge": true, "$out": true, "$facet": true,
}

// SanitizeFilter validates a query filter, returning a safe copy or an error.
// The original is not mutated.
func SanitizeFilter(filter M) (M, error) {
	if filter == nil {
		return M{}, nil
	}
	out, err := sanitizeDoc(filter, 0)
	if err != nil {
		return nil, err
	}
	return out.(M), nil
}

const maxFilterDepth = 12

func sanitizeDoc(v any, depth int) (any, error) {
	if depth > maxFilterDepth {
		return nil, fmt.Errorf("dopdb: filter nested too deeply (>%d)", maxFilterDepth)
	}
	switch t := v.(type) {
	case M: // note: M is an alias of map[string]any, so this also covers plain maps
		out := make(M, len(t))
		for k, val := range t {
			if err := checkKey(k); err != nil {
				return nil, err
			}
			sv, err := sanitizeDoc(val, depth+1)
			if err != nil {
				return nil, err
			}
			out[k] = sv
		}
		return out, nil
	case []any:
		out := make([]any, len(t))
		for i, e := range t {
			sv, err := sanitizeDoc(e, depth+1)
			if err != nil {
				return nil, err
			}
			out[i] = sv
		}
		return out, nil
	default:
		// scalar leaf (string/number/bool/nil/time/etc.) — safe as-is
		return v, nil
	}
}

func checkKey(k string) error {
	if !strings.HasPrefix(k, "$") {
		// A normal field path. Disallow operator dollar signs hidden mid-path
		// and the field-name injection vectors.
		if strings.Contains(k, "$") {
			return fmt.Errorf("dopdb: illegal field path %q", k)
		}
		return nil
	}
	if forbiddenOps[k] {
		return fmt.Errorf("dopdb: operator %q is not permitted", k)
	}
	if !allowedQueryOps[k] {
		return fmt.Errorf("dopdb: operator %q is not in the query allowlist", k)
	}
	return nil
}

File: ts/src/sanitize.ts
// Filter sanitization — a faithful TS port of sanitize.go.
//
// Mongo's query surface is an open document; exposing it raw to an untrusted
// caller is NoSQL injection by construction ({$where:"..."}, $function, cross
// collection $lookup, writes via $out/$merge, ...). SanitizeFilter walks a
// filter and admits ONLY a vetted operator allowlist, rejecting anything that
// can execute code, reach other collections, or write.
//
// This is the operator-level floor every Find passes through. Field-level
// scoping (which fields a collection exposes, the mandatory owner == @uid
// predicate) is layered on top by the server. Running it on the client too is
// a courtesy: bad filters fail before the network round-trip.

import { ValidationError } from "./errors.js";

export type Filter = Record<string, unknown>;

const ALLOWED = new Set<string>([
  // comparison
  "$eq", "$ne", "$gt", "$gte", "$lt", "$lte", "$in", "$nin",
  // logical
  "$and", "$or", "$nor", "$not",
  // element
  "$exists", "$type",
  // array
  "$all", "$elemMatch", "$size",
  // evaluation (safe subset)
  "$regex", "$options", "$mod",
]);

const FORBIDDEN = new Set<string>([
  "$where", "$function", "$accumulator",
  "$expr", // can embed $function/$let — disallow wholesale
  "$lookup", "$graphLookup", "$unionWith",
  "$merge", "$out", "$facet",
]);

const MAX_DEPTH = 12;

/** Validate a query filter, returning a safe copy. Throws ValidationError on a
 * forbidden operator, an unknown operator, or an illegal field path. The input
 * is not mutated. */
export function sanitizeFilter(filter: Filter | null | undefined): Filter {
  if (filter == null) return {};
  return sanitizeDoc(filter, 0) as Filter;
}

function sanitizeDoc(v: unknown, depth: number): unknown {
  if (depth > MAX_DEPTH) {
    throw new ValidationError([], `dopdb: filter nested too deeply (>${MAX_DEPTH})`);
  }
  if (Array.isArray(v)) {
    return v.map((e) => sanitizeDoc(e, depth + 1));
  }
  if (v !== null && typeof v === "object" && !(v instanceof Date)) {
    const out: Record<string, unknown> = {};
    for (const [k, val] of Object.entries(v as Record<string, unknown>)) {
      checkKey(k);
      out[k] = sanitizeDoc(val, depth + 1);
    }
    return out;
  }
  // scalar leaf (string/number/bool/null/Date) — safe as-is
  return v;
}

function checkKey(k: string): void {
  if (!k.startsWith("$")) {
    // a normal field path — reject dollar signs hidden mid-path
    if (k.includes("$")) {
      throw new ValidationError([], `dopdb: illegal field path "${k}"`);
    }
    return;
  }
  if (FORBIDDEN.has(k)) {
    throw new ValidationError([], `dopdb: operator "${k}" is not permitted`);
  }
  if (!ALLOWED.has(k)) {
    throw new ValidationError([], `dopdb: operator "${k}" is not in the query allowlist`);
  }
}
